### ISSUE TITLE
Replace apm-java-agent with OTEL SDK

### ISSUE THREAD

**Issue by pgomulka** [elastic/elasticsearch#109335]
### Description

elasticsearch is using apm-java-agent as the underlying implementation in the apm module.we are using our own apm api, implemented in apm-module with OTEL api. This should not change.
What should change is the binding between otel api and the implementation. Which should be otel sdk.Otel SDK will allow us to get more flexibility on configuring how our metrics and traces are sent to apm server (apm server support otel sdk). 
With Otel sdk we will be able to implement features like 'tee-ing' (splitting to two apm server) of the export or some additional buffering, retries when apm server is overloaded.

I worked on a simple very dirty PoC where this proves to work https://github.com/elastic/elasticsearch/compare/main...pgomulka:elasticsearch:otel_sdk_poc?expand=1
Things that need more investigation and work:

- [ ] 1. configuring the otel sdk with the apm-java-agent settings (things like metrics_interval, transactioN-sample_rate, server_url etc)
- [ ] 2. rework of RunTask - for localhost testing
- [ ] 3. rework of logging - otel sdk is using JUL logging. We already have a bridge in server https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/common/logging/JULBridge.java If we refactor it into a lib it should be possible to use it in the module. I wasn't able to make it work without adding a JUL -log4j bridge dependency (which we don't want) but at the same time I was rushing it.. We want to make sure we use the same apm_agent.json file (probably not worth renaming). This is a good example where plugin's might want to add an appender.
- [ ] 4. careful review of the dependencies. otel sdk requires quite a lot, some of the dependencies like okhttp are not working with java9's modules. some are introducing a 'clash version' dependencies (netty is already a dependency of server). It all shouldn't be a problem if we hide the implementation behind the embedded classloader like we do for x-content and jackson5. due to use of java beans the java9's module require a java.desktop. This feels awkward, but  I am not sure how to go around it.
- [ ] 6. OTEL sdk buildAndRegister can only be called once. If it is called twice and exception will be thrown. We need to make sure that starting/stopping the metering (this is possible now) will not throw this exception. 
- [ ] 7. apm-java-agent gives us a bunch of out of the box metrics for the jvm. I copied the JvmJdMetrics from apm-java-agent repo. Perhaps we need to work with apm team to have this as a lib? Just copying the JvmFdMetrics,  JvmGcMetrics, JvmMemoryMetrics could work initially, but feels dirty. The naming there has to also comply with our naming convention (we would register them using our with own api)
- [ ] 8. BIG - review the security manager permission. It would be a relatively tedious and long task, as there is a loot of new dependencies. For the PoC I have disabled the security manager

**Comment by pgomulka** [comment]
additional features to consider:
- being able to set custom interval for certain metrics
- change metric interval dynamically
- make sure metrics are sent upon node shutdown

**Comment by pgomulka** [comment]
we should also consider how to turn the exporting on and off

**Comment by SylvainJuge** [comment]
Hi !

For the JVM metrics, I am currently working on making them defined in the instrumentation as a yaml file (https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/13392) this won't be 100% equivalent to the ones captured by the OpenTelemetry instrumentation agent or EDOT due to limitations of the JMX interface, but would still be compliant with the [stable JVM metrics defined in semantic conventions](https://opentelemetry.io/docs/specs/semconv/runtime/jvm-metrics/).

With that, you should be able to easily reuse the `jmx-metrics` library module from instrumentation (it only depends on the SDK) and capture the JVM metrics in a semconv-compliant way. A good example of reusing this is the [jmx-scraper](https://github.com/open-telemetry/opentelemetry-java-contrib/tree/main/jmx-scraper) in contrib that is a CLI tool that connects to a remote JVM to extract metrics from the JMX interface and push them to an otel collector endpoint.

As an alternative, the [`runtime-telemetry/runtime-telemetry-java8/library`](https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/instrumentation/runtime-telemetry/runtime-telemetry-java8/library) (JMX) or [`runtime-telemetry/runtime-telemetry-java17`](https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/instrumentation/runtime-telemetry/runtime-telemetry-java17/library) (JMX + JFR) library modules in instrumentation could also provide an easier setup without the limitations of the JMX + YAML implementation and they are available today.

Something worth considering is that the metrics mapping might be a bit different from what they currently do, which means having potentially breaking changes on anything that relies on them (dashboards, alerts, ...).

In addition, if we have to deal with breaking changes here, it's probably a good idea to bundle this with changing the ingestion path to use the EDOT collector instead of the apm endpoint. Doing so should bring a more otel-native storage of data while preserving compatibility with APM UI.

**Comment by JonasKunz** [comment]
> 10. otel sdk exporters support http and grpc protocoles. APM server works with both. Need to decide on one.

The default exporter is `http/protobuf`, due to the easier, more widespread support than gRPC. For the detailed reasoning please see this [issue](https://github.com/open-telemetry/opentelemetry-specification/issues/1885).
Therefore I'd suggest to stick with http, as it is the default.

**Comment by JonasKunz** [comment]
[Here](https://github.com/JonasKunz/otel-sdk-dynamic-demo)'s a demo project showing how to:
 * Create and globally register an OpenTelemetrySDK instance
 * Have the exporters dynamically configurable on that instance

If you need any help with understanding the demo or anything else let me know!

**Comment by mosche** [comment]
Health metrics similar to the ones published by the agent should be another consideration here.

https://www.elastic.co/docs/reference/apm/agents/java/metrics#metrics-agenthealth

**Comment by JonasKunz** [comment]
We actually put great effort into getting health metrics standardized and implemented in OTel:
https://opentelemetry.io/docs/specs/semconv/otel/sdk-metrics/

The most important ones to my knowledge for the ES use case (=health metrics on exporters) are already in the latest Otel SDK release

**Comment by mosche** [comment]
Another long outstanding issue we should revisit is missing counter metrics when counters are rarely incremented.
Due to dynamic mappings and frequent rollovers such metrics might not be known at all to Kibana, making them difficult to be used on dashboards. (For context, counters are only lazily reported when the delta is > 0.)

See also https://github.com/elastic/observability-dev/issues/3042

**Comment by felixbarny** [comment]
> Due to dynamic mappings and frequent rollovers such metrics might not be known at all to Kibana, making them difficult to be used on dashboards.

This seems related to [supporting optional fields in ES|QL](https://github.com/elastic/elasticsearch/issues/112912).

**Comment by prdoyle** [comment]
I'm going to capture a snapshot of this PR's description as it is right now. Work on this is picking up again and we might find it useful to change the description.

-------------------

### Description

elasticsearch is using apm-java-agent as the underlying implementation in the apm module.we are using our own apm api, implemented in apm-module with OTEL api. This should not change.
What should change is the binding between otel api and the implementation. Which should be otel sdk. Otel SDK will allow us to get more flexibility on configuring how our metrics and traces are sent to apm server (apm server support otel sdk). 
With Otel sdk we will be able to implement features like 'tee-ing' (splitting to two apm server) of the export or some additional buffering, retries when apm server is overloaded.

I worked on a simple very dirty PoC where this proves to work https://github.com/elastic/elasticsearch/pull/110263
Things that need more investigation and work:

- [ ] 1. configuring the otel sdk with the apm-java-agent settings (things like metrics_interval, transactioN-sample_rate, server_url etc)
- [ ] 2. rework of RunTask - for localhost testing https://github.com/elastic/elasticsearch/pull/110263/commits/73c0e5f1fd137ba46452153fb565f1864731194e
- [x] ~3. rework of logging - otel sdk is using JUL logging. We already have a bridge in server https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/common/logging/JULBridge.java If we refactor it into a lib it should be possible to use it in the module. I wasn't able to make it work without adding a JUL -log4j bridge dependency (which we don't want) but at the same time I was rushing it.. We want to make sure we use the same apm_agent.json file (probably not worth renaming). This is a good example where plugin's might want to add an appender.~
- [x] 4. careful review of the dependencies. otel sdk requires quite a lot, some of the dependencies like okhttp are not working with java9's modules. some are introducing a 'clash version' dependencies (netty is already a dependency of server). It all shouldn't be a problem if we hide the implementation behind the embedded classloader like we do for x-content and jackson
- [ ] 5. due to use of java beans the java9's module require a java.desktop. This feels awkward, but  I am not sure how to go around it.
- [ ] 6. OTEL sdk buildAndRegister can only be called once. If it is called twice and exception will be thrown. We need to make sure that starting/stopping the metering (this is possible now) will not throw this exception. 
- [ ] 7. apm-java-agent gives us a bunch of out of the box metrics for the jvm. I copied the JvmJdMetrics from apm-java-agent repo. Perhaps we need to work with apm team to have this as a lib? Just copying the JvmFdMetrics,  JvmGcMetrics, JvmMemoryMetrics could work initially, but feels dirty. The naming there has to also comply with our naming convention (we would register them using our with own api)
- [ ] 8. BIG - review the security manager permission. It would be a relatively tedious and long task, as there is a loot of new dependencies. For the PoC I have disabled the security manager
- [ ] 9. New Exporters - we could simply configure the out of the box available exporters (simply adding 2 for the support of exporting to 2 apm servers) or implement our own so that we have more control of logging etc
- [ ] 10. otel sdk exporters support http and grpc protocoles. APM server works with both. Need to decide on one.

**Comment by SylvainJuge** [comment]
Switching to OTel SDK also means that changes to how the metrics are processed/stored:
- the metrics will be sent by the OTel SDK through the OTLP protocol
- those metrics can be mapped and stored to either "ECS" or "OTel native" formats, with APM Agents it was only the "ECS" format.

While it is possible to use apm-server or the otel collector to attempt preserving the mapping, switching to an "otel native" storage is the most future-proof.

As a consequence, this means that we have to deal with potential breaking changes on those metrics, as their "shape" might change a bit.

At the same time, making Elasticsearch produce OpenTelemetry metrics that are compatible with [semantic conventions](https://opentelemetry.io/docs/specs/semconv/) is (in my opinion) essential to make ES a "good citizen" of the OpenTelemetry ecosystem and avoid re-inventing the wheel when we can.

This includes attempting to achieve the following:
- reuse existing semantic conventions when applicable: this includes for example JVM metrics
- attempt to make Elasticsearch metrics compatible with [semconv recommendations for metrics](https://opentelemetry.io/docs/specs/semconv/general/metrics/), this implies revisiting most of the [current metric naming guide](https://github.com/elastic/elasticsearch/blob/main/modules/apm/NAMING.md), for example:
  - namespace metric with product name `elasticsearch.*`
  - namespace metric attributes with product name `elasticsearch.*`
  - avoid `.total` as suffix
  - use standard units, or singular nouns as units, for example `{request}`

It is probably quite hard to bundle SDK change + revisit all metrics in a single step, and this must be done in multiple intermediate steps.

To summarize:
- using the OTel SDK is likely to be a breaking change as metrics will not be stored exactly as they used to
- aligning metrics to be consistent with semconv is something that will also introduce breaking changes
- we need to be able to deal with those breaking changes for the consumers of those metrics, also semantic conventions might change over time

**Comment by mosche** [comment]
> those metrics can be mapped and stored to either "ECS" or "OTel native" formats, with APM Agents it was only the "ECS" format.

@SylvainJuge what's the implications of using OTel native formats? One key requirement of the migration certainly is to not break Serverless dashboards and alerts.

> - attempt to make Elasticsearch metrics compatible with [semconv recommendations for metrics](https://opentelemetry.io/docs/specs/semconv/general/metrics/), this implies revisiting most of the [current metric naming guide](https://github.com/elastic/elasticsearch/blob/main/modules/apm/NAMING.md), for example:
>   - namespace metric with product name elasticsearch.*
>   - namespace metric attributes with product name elasticsearch.*

The current product namespace for metrics and attributes is simply `es.*`, imho we cannot afford changing that anymore.

Could you elaborate on the other recommendations / feedback on the naming guide? I was hoping that was initially developed in collaboration with your team, but it doesn't sound like 😞

**Comment by SylvainJuge** [comment]
> @SylvainJuge what's the implications of using OTel native formats? One key requirement of the migration certainly is to not break Serverless dashboards and alerts.

I'm definitely not an expert on that part of ingestion, but if the metric names and attributes remain the same, then most of things _should_ still work as they do, however:
- there are a few differences between the two data streams : https://www.elastic.co/docs/reference/opentelemetry/data-streams
- there might be a few differences in field types
- it might be simpler to test and see if there are breaking changes.

> The current product namespace for metrics and attributes is simply es.*, imho we cannot afford changing that anymore.

I understand we might not afford breaking that for the sake of what can appear as cosmetic changes, as long as this is for internal use this seems acceptable to me.

However, if those metrics are meant for end-users, they should be properly specified and documented and any user could reuse them, for example with [MySQL OpenTelemetry metrics](https://dev.mysql.com/doc/refman/8.4/en/telemetry-metrics.html).

I've also asked for guidance in the OTel [semconv slack here](https://cloud-native.slack.com/archives/C041APFBYQP/p1768932848622279), and the summary is that:
- there isn't any strong requirement for products to fit into semantic conventions (but I think it would still be beneficial).
- metrics/attributes should be documented
- metrics/attributes should be stable

Also, do we have an exhaustive list of all the metrics that are captured or do we have to extract this from the code as reference ?

Maybe here we should keep focused on the OTel SDK migration first, then maybe attempt to rework the metrics definitions if we have a good incentive to do so.

**Comment by SylvainJuge** [comment]
Also, in the case of JVM metrics, reusing the definitions from semantic conventions will introduce some breaking changes because the metric names and attributes are just not the same.
We can use that as a way to evaluate the implications of modifying those metrics.

However, this also brings a few benefits like the ability to reuse dashboards/visualizations and content that may be available for JVMs.

**Comment by mosche** [comment]
> However, if those metrics are meant for end-users, they should be properly specified and documented and any user could reuse them, for example with [MySQL OpenTelemetry metrics](https://dev.mysql.com/doc/refman/8.4/en/telemetry-metrics.html).

This is a bit of a controversial point I admit, while we consider supporting multiple exporters to report metrics data to customer clusters in ECH as well, we do not intend to establish any contracts on metrics currently and reserve the right to make changes. This is something we should revisit at some point, but shouldn't impact this migration project. 

> Also, do we have an exhaustive list of all the metrics that are captured or do we have to extract this from the code as reference ?

There's a logger, but that's certainly an area that can/should be improved
./gradlew run -Dtests.es.logger.org.elasticsearch.telemetry.apm=debug

**Comment by SylvainJuge** [comment]
Even if we just start with the JVM runtime metrics, I think it would be beneficial for us to document the changes to make the migration easier, I've opened https://github.com/elastic/elastic-otel-java/issues/960 to take care of this.

Also, we aim to document this to target the "otel native" data storage format and assume a recent version of ES stack (ideally Serverless) is used to consume those metrics, are there any constraints on the ES team side about this ?

While this would not help solve the "which metrics are used in any of the dashboards/alerts/...", this would at least provide hands-on guidance on how to deal with the changes (for example by providing alternative metrics or ESQL queries to get equivalent information).
